pymtl3/datatypes/bit_structs.py

```diff
@@ -0,0 +1,442 @@
+# Creates a init argument string from a field.
+#
+# Notes:
+#
+# We store the default value in _globals by adding a _deflt_ prefix to the
+# field name. This seems dangerous as our Bits and bit structs objects are
+# all mutable. Another idea is to use repr to re-construct a new instance
+# but that seems to be too much to ask.
+
+def _mk_init_arg( f ):
+
+  # Call the default constructor if no default value specified
+  if f.default is MISSING:
+    default = f'{f.type_.__name__}()'
+
+  # Use the default value if it is specified by user
+  else:
+    default = f'_deflt_{f.name}'
+
+  # return f'{f.name} : _type_{f.name} = {default}'
+  return f'{f.name} : {f.type_.__name__} = {default}'
+
+#-------------------------------------------------------------------------
+# _mk_init_body
+#-------------------------------------------------------------------------
+# Creates one line of __init__ body from a field and add its default value
+# to globals.
+
+def _mk_init_body( self_name, f ):
+  return f'{self_name}.{f.name} = {f.name}'
+
+#-------------------------------------------------------------------------
+# _mk_tuple_str
+#-------------------------------------------------------------------------
+# Creates a tuple of string representations of each field. For example,
+# if the self_name is 'self' and fields is [ 'x', 'y' ], it will return
+# ('self.x', 'self.y'). This is used for creating the default __eq__ and
+# __hash__ function.
+
+def _mk_tuple_str( self_name, fields ):
+  return f'({",".join([f"{self_name}.{f.name}" for f in fields])},)'
+
+#-------------------------------------------------------------------------
+# _mk_init_fn
+#-------------------------------------------------------------------------
+# Creates a __init__ function based on fields.
+# TODO: should we provide a __post_init__ function like dataclass does?
+
+def _mk_init_fn( self_name, fields ):
+
+  # Sanity check: make sure no postional argument after keyword argument
+  seen_kwarg = False
+  for f in fields:
+    if not f.default is MISSING:
+      seen_kwarg = True
+    elif seen_kwarg:
+      raise TypeError( f'Non keyword argument {f.name} '
+                       'follows keyword argument' )
+
+  _globals = { 'MISSING' : MISSING }
+
+  # Register default values to _globals
+  for f in fields:
+    if not (f.default is MISSING):
+      _globals[f'_deflt_{f.name}'] = f.default
+
+  # Register types in _locals
+  _locals = { f'{f.type_.__name__}' : f.type_ for f in fields }
+
+  return _create_fn(
+    '__init__',
+    [ self_name ] + [ _mk_init_arg( f ) for f in fields ],
+    [ _mk_init_body( self_name, f ) for f in fields ],
+    _locals  = _locals,
+    _globals = _globals,
+  )
+
+#-------------------------------------------------------------------------
+# _mk_str_fn
+#-------------------------------------------------------------------------
+# Creates a __str__ function based on fields.
+
```

Comment:
Since you are generating some code from a list here, can you provide one example for each function on the actual code it generates?